plugins/cros-ec/fu-cros-ec-usb-device.c

```diff
@@ -299,6 +314,248 @@ fu_cros_ec_usb_device_setup (FuDevice *device, GError **error)
 	return TRUE;
 }
 
+static gboolean
+fu_cros_ec_usb_device_transfer_block (FuDevice *device, gpointer user_data,
+				      GError **error)
+{
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	fu_cros_ec_usb_block_info *block_info = (fu_cros_ec_usb_block_info *) user_data;
+	gsize image_size = 0;
+	gsize transfer_size = 0;
+	guint32 reply = 0;
+	g_autoptr(GBytes) block_bytes = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	if (block_info == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "no block_info provided");
+		return FALSE;
+	}
+
+	image_size = g_bytes_get_size (block_info->image_bytes);
+	if (block_info->offset + block_info->payload_size > image_size) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "offset %" G_GSIZE_FORMAT "plus payload_size %"
+			     G_GSIZE_FORMAT " exceeds image size %"
+			     G_GSIZE_FORMAT,
+			     block_info->offset, block_info->payload_size,
+			     image_size);
+		return FALSE;
+	}
+
+	block_bytes = g_bytes_new_from_bytes (block_info->image_bytes,
+					      block_info->offset,
+					      block_info->payload_size);
+	chunks = fu_chunk_array_new_from_bytes (block_bytes,
+						0x00,
+						0x00,
+						self->chunk_len);
+
+	/* first send the header */
+	if (!fu_cros_ec_usb_device_do_xfer (self, (guint8 *)&block_info->ufh,
+					    sizeof(struct update_frame_header),
+					    NULL,
+					    0, FALSE,
+					    NULL, error))
+		return FALSE;
+
+	/* send the block, chunk by chunk */
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+
+		if (!fu_cros_ec_usb_device_do_xfer (self,
+						    (guint8 *)chk->data,
+						    chk->data_sz,
+						    NULL,
+						    0, FALSE,
+						    NULL, error)) {
+			return FALSE;
+		}
+	}
+
+	/* get the reply */
+	if (!fu_cros_ec_usb_device_do_xfer (self, NULL, 0,
+					    (guint8 *)&reply, sizeof (reply),
+					    TRUE, &transfer_size, error))
+		return FALSE;
+	if (transfer_size == 0) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "zero bytes received for block reply");
+		return FALSE;
+	}
+	if (reply != 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "error: status 0x%#x", reply);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_transfer_section (FuDevice *device,
+					FuFirmware *firmware,
+					FuCrosEcFirmwareSection *section,
+					gboolean smart_update,
```

Comment:
It seems to me that you only use this function in one place and always send `smart_update` as `TRUE`.  Worth dropping?

Reply:
Done.